File: Data-Structures-and-Algorithms/1047-RemoveAllAdjacentDuplicatesinString/removeDuplicates.py
# Solution using hashmap
from string import ascii_lowercase
class Solution(object):
    def removeDuplicates(self, s):
        # making a hashmap of all lowercase characters
        duplicates = {2 * ch for ch in ascii_lowercase}
        
        prev_length = -1
        while prev_length != len(s):
            prev_length = len(s)
            for d in duplicates:
                s = s.replace(d, '')
        return s

S = 'aabbccioppoil'
print(Solution().removeDuplicates(S))

# O(N^2) for time, because we traverse N times with replace function in each N
# O(N) for space, becasue replace function keeps a copy
# Remove all adjacent repeating elements

File: Data-Structures-and-Algorithms/1047-RemoveAllAdjacentDuplicatesinString/removeDuplicates.js
// Solution using stack track
const removeDuplicates = function(s) {
    // initiate the resulting array of characters to be joined
    const res = [];
    // for each character of string s
    for (var ch of s) {  
        if (ch == res[res.length-1]) {
            res.pop()
        } else {
            res.push(ch)
        }
    }
    // join the array elements with nothing inbetween
    return res.join('')
};

S = 'aabbccioppoilaabbpweiqwoebcasjjjdal'
console.log(removeDuplicates(S))

// O(N) for time, because we traverse once through the string of length N
// O(N) for space, we store the res as array then join
// Remove all adjacent repeating elements